include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,808 @@
+
+      AssertIndexRange(cell_index - start, ptrs_ptrs.size());
+      const unsigned int face_index =
+        ptrs_ptrs[cell_index - start] + face_number;
+      AssertIndexRange(face_index, ptrs.size());
+      return ptrs[face_index];
+    }
+
+    /**
+     * Get the number of stored values.
+     */
+    unsigned int
+    size() const
+    {
+      Assert(ptrs.size() > 0, ExcInternalError());
+      return ptrs.back();
+    }
+
+    /**
+     * This parameter can be used if indices do not start with 0.
+     */
+    unsigned int start = 0;
+
+    /**
+     * Pointers to @c ptrs.
+     */
+    std::vector<unsigned int> ptrs_ptrs;
+
+    /**
+     * Pointers to data.
+     */
+    std::vector<unsigned int> ptrs;
+  };
+
+  /**
+   * A class helps to access @c PrecomputedEvaluationData in a thread safe
+   * manner. Each thread has to create this wrapper class on its own to
+   * avoid race-conditions during @c reinit().
+   */
+  template <int dim, int n_components, typename value_type_>
+  class PrecomputedEvaluationDataAccessor
+  {
+    using value_type =
+      typename PrecomputedEvaluationData<dim, n_components, value_type_>::
+        value_type;
+    using gradient_type =
+      typename PrecomputedEvaluationData<dim, n_components, value_type_>::
+        gradient_type;
+
+  public:
+    PrecomputedEvaluationDataAccessor(
+      const PrecomputedEvaluationData<dim, n_components, value_type_> &data,
+      const PrecomputedEvaluationDataView                             &view)
+      : view(view)
+      , data(data)
+      , data_offset(numbers::invalid_unsigned_int)
+    {}
+
+    /**
+     * Get the value at quadrature point @p q. The entity on which the values
+     * are defined is set via `reinit()`.
+     */
+    const value_type
+    get_value(const unsigned int q) const
+    {
+      Assert(data_offset != numbers::invalid_unsigned_int,
+             ExcMessage("reinit() not called."));
+      AssertIndexRange(data_offset + q, data.values.size());
+      return data.values[data_offset + q];
+    }
+
+    /**
+     * Get the gradients at quadrature point @p q. The entity on which the
+     * gradients are defined is set via `reinit()`.
+     */
+    const gradient_type
+    get_gradient(const unsigned int q) const
+    {
+      Assert(data_offset != numbers::invalid_unsigned_int,
+             ExcMessage("reinit() not called."));
+      AssertIndexRange(data_offset + q, data.gradients.size());
+      return data.gradients[data_offset + q];
+    }
+
+
+    /**
+     * Set entity index at which quadrature points are accessed. This can, e.g.,
+     * a cell index, a cell batch index, or a face batch index.
+     */
+    void
+    reinit(const unsigned int index)
+    {
+      data_offset = view.get_shift(index);
+    }
+
+    /**
+     * Set cell and face_number at which quadrature points are accessed.
+     */
+    void
+    reinit(const unsigned int cell_index, const unsigned int face_number)
```

Comment:
Let's keep this general.
```suggestion
    reinit(const unsigned int index_0, const unsigned int index_1)
```